[INTEL MKL]Optimize CropAndResizeGradImage Op.

pandaoxin opened the pull request:
Optimize CropAndResizeGradImage Op, please review. After optimized, the performance could raise about  20.6% compared to original version. We use parallization(eigen thread pool) to optimize.

pandaoxin committed:
Optimize CropAndResizeGradImage Op.

qlzh727 commented:
let me rerun the build, the previous one seems to be failed due to build system issue.

pandaoxin commented:
Any updates? @qlzh727 @rmlarsen

pandaoxin commented:
Any updates? @qlzh727 @rmlarsen

pandaoxin commented:
Any updates? @qlzh727 @rmlarsen @agramesh1

pandaoxin commented:
Any updates? @qlzh727 @rmlarsen

pandaoxin commented:
Any updates? @qlzh727 @rmlarsen

qlzh727 commented:
@rmlarsen is on his holiday until next week. Adding @shpeisman for review.

pandaoxin commented:
@qlzh727 Thank you!

qlzh727 commented:
Ping @tatianashp for review. Thanks.

pandaoxin commented:
@qlzh727 Thank you~

pandaoxin commented:
Any updates? @qlzh727 @tatianashp

qlzh727 commented:
Sorry for the very late reply, I have pinged @tatianashp and she will probably reply very soon.

pandaoxin commented:
Thank you @qlzh727

tensorflowbutler commented:
Nagging Reviewer @rmlarsen, @tatianashp: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 15 days with no activity and the `awaiting review` label has been applied.

pandaoxin commented:
Any updates?

qlzh727 commented:
Sorry for the very late reply, I will ping reviewers again today for this PR. In the meantime, can you resolve the merge conflicts?

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -442,70 +455,96 @@ struct CropAndResizeBackpropImage<CPUDevice, T> {
+    const double cost_per_pixel =
+        (isBilinear
+             ? depth * (Eigen::TensorOpCost::AddCost<float>() * 4 +
+                        Eigen::TensorOpCost::MulCost<float>() * 4 +
```
I see 6 multiplications in there?
```suggestion
                        Eigen::TensorOpCost::MulCost<float>() * 6 +
```

pandaoxin replied:
Done.

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -442,70 +455,96 @@ struct CropAndResizeBackpropImage<CPUDevice, T> {
+          const int top_y_index = floorf(in_y);
+          const int bottom_y_index = ceilf(in_y);
+          const float y_lerp = in_y - top_y_index;
+          const float one_y_lerp = 1 - y_lerp;
```
`top_y_index`, `bottom_y_index`, `y_lerp`, and `one_y_lerp` are only used in `bilinear`. Please move them into `if (isBilinear)`.

pandaoxin replied:
Done.

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -442,70 +455,96 @@ struct CropAndResizeBackpropImage<CPUDevice, T> {
-                  static_cast<T>((1 - x_lerp) * dbottom);
-              grads_image(b_in, bottom_y_index, right_x_index, d) +=
-                  static_cast<T>(x_lerp * dbottom);
+          const int closest_y_index = roundf(in_y);
```
`closest_y_index` and `closest_x_index` are only used in `nearest`. Please move them into `else`, before the `depth` loop.

pandaoxin replied:
Done.

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -442,70 +455,96 @@ struct CropAndResizeBackpropImage<CPUDevice, T> {
+             : depth * (Eigen::TensorOpCost::AddCost<float>() +
+                        Eigen::TensorOpCost::CastCost<T, float>())) +
+        (Eigen::TensorOpCost::AddCost<float>() * 2 +
+         Eigen::TensorOpCost::AddCost<float>() * 3);
```
Let's update these out-of-depth-loop costs after we are done moving the instructions around.

pandaoxin replied:
OK!

penpornk replied:
It looks like we have
- 4 additions and no multiplications out of the`depth` loop for `bilinear`, and
- no extra additions/multiplications out of the `depth` loop for `nearest`.
Please update the costs. :)

pandaoxin replied:
It's done.

penpornk reviewed (changes requested):
Thank you for the PR! Please resolve the merge conflicts.

pandaoxin committed:
Merged from master.

pandaoxin committed:
Some changes.

pandaoxin commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -423,14 +435,15 @@ class CropAndResizeGradImageOp : public AsyncOpKernel {
-template <typename T>
-struct CropAndResizeBackpropImage<CPUDevice, T> {
-  bool operator()(const CPUDevice& d,
+// Split two methods' code to avoid branch mis-prediction.
```
Original branch locates in the loop, so I think current situation will be more efficient. Maybe more tests could be explored.

penpornk replied:
You're right. I forgot about the loops. I still don't like that we have to have `if (isBilinear)` in there, even though the branch mispredictions will probably be low. How about just writing two lambda functions and storing the right one in a variable before entering the `b` for loop (and not add a third template parameter)?

penpornk replied:
You're right. I forgot about the loops. I still don't like that we have to have `if (isBilinear)` in there, even though the branch mispredictions will probably be low. How about just writing two lambda functions and storing the right one in a variable before entering the `b` for loop (and not add a third template parameter)? What do you think?

pandaoxin committed:
Modify calculation costs.

pandaoxin committed:
Change function splitting method.

pandaoxin commented:
@penpornk Your idea is good. For the 'two lambda functions', do you mean by the implementation in my latest commit?

pandaoxin committed:
Clang format issues.

penpornk reviewed (changes requested):
@pandaoxin That works too. But both method share some code, I was thinking more of factoring just the code inside the if. For example, defining `resize_fn`
```c++
auto resize_fn = (method_name == "bilinear")
    ? <lambda function for "bilinear">
    : <lambda function for "nearest">;
```
before the for loop 
https://github.com/tensorflow/tensorflow/blob/81fdde74edab29ecde35e2f4471531d1d95f2b96/tensorflow/core/kernels/crop_and_resize_op.cc#L445

Where the lambda function for `bilinear` contains
https://github.com/tensorflow/tensorflow/blob/81fdde74edab29ecde35e2f4471531d1d95f2b96/tensorflow/core/kernels/crop_and_resize_op.cc#L482-L496
And the lambda function for `nearest` contains
https://github.com/tensorflow/tensorflow/blob/81fdde74edab29ecde35e2f4471531d1d95f2b96/tensorflow/core/kernels/crop_and_resize_op.cc#L499-L504

Finally, inside the for loop, we replace all this
https://github.com/tensorflow/tensorflow/blob/81fdde74edab29ecde35e2f4471531d1d95f2b96/tensorflow/core/kernels/crop_and_resize_op.cc#L481-L505
with just one call to `resize_fn`.

What do you think?

penpornk commented:
Quoting myself to add:
> For example, defining `resize_fn`
> 
> ```c++
> auto resize_fn = (method_name == "bilinear")
>     ? <lambda function for "bilinear">
>     : <lambda function for "nearest">;
> ```
Since each lambda takes multiple lines, you might want to define `resize_fn` this way instead.
```c++
auto resize_fn = <lambda_function_for_bilinear>;
if (method_name == "nearest") {
  resize_fn = <lambda_function_for_nearest>;
}
```

pandaoxin committed:
Further modification.

pandaoxin committed:
Fix format issues.

pandaoxin commented:
@penpornk Thank you! Changes done.

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -216,7 +217,8 @@ struct CropAndResize<CPUDevice, T> {
     const int depth = crops.dimension(3);
 
     // Sharding across boxes.
-    auto CropAndResizePerBox = [&](int start_box, int limit_box) {
```
I meant making 2 lambda functions to be called inside `CropAndResizePerBox`, like this. Sorry for not being clear earlier!
```c++
    // Bilinear interpolation.
    auto interpolation_fn = [&](const float in_y) {
      const int top_y_index = floorf(in_y);
      const int bottom_y_index = ceilf(in_y);
      const float y_lerp = in_y - top_y_index;

      for (int x = 0; x < crop_width; ++x) {
        const float in_x = (crop_width > 1)
                               ? x1 * (image_width - 1) + x * width_scale
                               : 0.5 * (x1 + x2) * (image_width - 1);
        if (in_x < 0 || in_x > image_width - 1) {
          for (int d = 0; d < depth; ++d) {
            crops(b, y, x, d) = extrapolation_value;
          }
          continue;
        }
        const int left_x_index = floorf(in_x);
        const int right_x_index = ceilf(in_x);
        const float x_lerp = in_x - left_x_index;

        for (int d = 0; d < depth; ++d) {
          const float top_left(static_cast<float>(
              image(b_in, top_y_index, left_x_index, d)));
          const float top_right(static_cast<float>(
              image(b_in, top_y_index, right_x_index, d)));
          const float bottom_left(static_cast<float>(
              image(b_in, bottom_y_index, left_x_index, d)));
          const float bottom_right(static_cast<float>(
              image(b_in, bottom_y_index, right_x_index, d)));
          const float top = top_left + (top_right - top_left) * x_lerp;
          const float bottom =
              bottom_left + (bottom_right - bottom_left) * x_lerp;
          crops(b, y, x, d) = top + (bottom - top) * y_lerp;
        }
      }
    };
    if (method == "nearest") {  // Nearest-neighbor interpolation.
      interpolation_fn = [&](const float in_y) {
        for (int x = 0; x < crop_width; ++x) {
          const float in_x = (crop_width > 1)
                                 ? x1 * (image_width - 1) + x * width_scale
                                 : 0.5 * (x1 + x2) * (image_width - 1);
          if (in_x < 0 || in_x > image_width - 1) {
            for (int d = 0; d < depth; ++d) {
              crops(b, y, x, d) = extrapolation_value;
            }
            continue;
          }
          const int closest_x_index = roundf(in_x);
          const int closest_y_index = roundf(in_y);
          for (int d = 0; d < depth; ++d) {
            crops(b, y, x, d) = static_cast<float>(
                image(b_in, closest_y_index, closest_x_index, d));
          }
        }
      };
    }
    auto CropAndResizePerBox = [&](int start_box, int limit_box) {
      for (int b = start_box; b < limit_box; ++b) {
        const float y1 = boxes(b, 0);
        const float x1 = boxes(b, 1);
        const float y2 = boxes(b, 2);
        const float x2 = boxes(b, 3);

        const int32 b_in = box_index(b);
        if (!FastBoundsCheck(b_in, batch_size)) {
          continue;
        }

        const float height_scale =
            (crop_height > 1)
                ? (y2 - y1) * (image_height - 1) / (crop_height - 1)
                : 0;
        const float width_scale =
            (crop_width > 1) ? (x2 - x1) * (image_width - 1) / (crop_width - 1)
                             : 0;

        for (int y = 0; y < crop_height; ++y) {
          const float in_y = (crop_height > 1)
                                 ? y1 * (image_height - 1) + y * height_scale
                                 : 0.5 * (y1 + y2) * (image_height - 1);
          if (in_y < 0 || in_y > image_height - 1) {
            for (int x = 0; x < crop_width; ++x) {
              for (int d = 0; d < depth; ++d) {
                crops(b, y, x, d) = extrapolation_value;
              }
            }
            continue;
          }
          interpolation_fn(in_y);
        }
      }
    };
```

pandaoxin committed:
More changes.

pandaoxin committed:
Fix format.

pandaoxin commented:
@penpornk No problem! Changes done.

wenxizhu commented:
Hi @penpornk @ezhulenev 

@pandaoxin will leave his job this week and start a new career. From now on, I will take over this PR and until it's merged.

Is there anything I should do to make the ownership transfer of this PR more smoothly? For example,  re-check CLA? If so, please let me know.

Thank you!

penpornk commented:
@pandaoxin Sorry for my late response. I wish you all the best on your new journey!

@wenxizhu Since you have committed to the repo before, I believe your CLA is fine. But googlebot will probably complain that there is more than one users contributing to the PR. And both @pandaoxin and you will have to post that you are okay with your commits being used in this PR.

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -216,7 +217,73 @@ struct CropAndResize<CPUDevice, T> {
 
     // Sharding across boxes.
-    auto CropAndResizePerBox = [&](int start_box, int limit_box) {
+    std::function<void(int32 b_in, int b, int y, const float x1, const float x2,
```
Please avoid storing lambda functions in `std::function`. I don't think `std::function` can be inlined, which will make the whole thing slower. Please use `auto` instead. The same goes for other lambdas in this PR. Thank you!

wenxizhu replied:
I'm afraid we can't simply use `auto` to replace `std::function`, because a lambda function is const thus can't be assign to a new value in the following 'nearest' case line.

wenxizhu replied:
I propose following 2 alternative approaches, if you insist to use lambda functions:
1. Change the whole structure of the wrapper function to:  
`if (method_name == "bilinear") {
         auto interpolation_fn = [&](){...}
         auto CropAndResizeBackImagePerBox = [&](){...}
         // A bunch of code
         Shard(worker_threads.num_threads...
         return true;
     else {
         auto interpolation_fn = [&](){...}
         auto CropAndResizeBackImagePerBox = [&](){...}
         // A bunch of code
         Shard(worker_threads.num_threads...
         return true;
     }
 `

penpornk replied:
We can use a ternary operator 
```
auto lambda = (method_name == "bilinear ? method_a : method_b;
```

But I totally agree with your 2. Let's do 2. Thank you! :)

wenxizhu replied:
2 applied! We can move on now.

wenxizhu commented:
@penpornk I'm sorry @pandaoxin already left his job yesterday, so he won't be able to login his account (it's a company account) to post that comment. Is there any way we can workaround this?

penpornk commented:
@wenxizhu Sorry for the inconvenience! Let's wait until googlebot complains and I'll ask someone. At worst we'll just open a new PR.

wenxizhu committed:
Revert the change of use "interpolation_fn".

We introduced "interpolation_fn" to avoid pipeline stall caused by branch
mis-prediction. However we revisited the code and confirmed that won't happen,
so we revert the change now.

wenxizhu commented:
@penpornk Changes applied. Any updates?

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -442,70 +443,92 @@ struct CropAndResizeBackpropImage<CPUDevice, T> {
+             ? depth * (Eigen::TensorOpCost::AddCost<float>() * 4 +
+                        Eigen::TensorOpCost::MulCost<float>() * 6 +
+                        Eigen::TensorOpCost::CastCost<T, float>() * 4) +
+                   Eigen::TensorOpCost::AddCost<float>() * 4
```
I only found three additions, `in_x - left_x_index`, `1 - y_lerp` , and `1 - x_lerp`. Where is the 4th one?

wenxizhu replied:
My best guess is:
1. line 467:
    <pre>
    const float in_y = (crop_height > 1)
                                 ? y1 * (image_height - 1) + y * height_scale
                                 : 0.5 * (y1 + y2) * (image_height - 1);
    </pre>
    both branch has 2 additions;
2. line 481:
    <pre>
    if (in_x < 0 || in_x > image_width - 1) {
    </pre>
    That's 1 addition;
3. line 488:
    <pre>
    const float x_lerp = in_x - left_x_index;
    </pre>
    That's 1 addition.

wenxizhu replied:
And I suppose you count `1 - y_lerp` and `1 - x_lerp` as non-loop additions because you assume compiler will do loop invariant motion for them? But don't we not consider compiler optimizations in cost model, because cost model is merely a rough estimation? I don't know.

penpornk replied:
So sorry for my late reply! I only looked for addition in the `bilinear` if case because you only added 4 `AddCost`s to the `bilinear` cost and not to `nearest`. In the `if` there are 4 `+=` (at the depth level) so I guessed you counted `1 - y_lerp` and `1 - x_lerp` as loop invariants. I don't usually try to count for compiler optimizations either. :) In the end this is just a rough estimate to help shard the work better. I agree about ignoring optimizations.

I didn't author this code too. How about listing the costs explicitly here and see if there is anything missing? I'll start:

Cost at the pixel level (both methods):
2 additions and 2 multiplications for either branch.
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L478-L480
1 addition from
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L481

Cost at the pixel level for just `bilinear`:
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L488

Cost at the depth level for `bilinear`:
3 additions for `1 - y_lerp` and `1 - x_lerp`
4 additions for `+=`
6 multiplications

Cost at the depth level for `nearest`
1 addition and 1 cast.

Anything else?

penpornk replied:
I only looked for addition in the `bilinear` if case because you only added 4 `AddCost`s to the `bilinear` cost and not to `nearest`. In the `if` there are 4 `+=` (at the depth level) so I guessed you counted `1 - y_lerp` and `1 - x_lerp` as loop invariants. I don't usually try to count for compiler optimizations either. :) In the end this is just a rough estimate to help shard the work better. I agree about ignoring optimizations.

I didn't author this code too. How about listing the costs explicitly here and see if there is anything missing? I'll start:

Cost at the pixel level (both methods):
2 additions and 2 multiplications for either branch.
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L478-L480
1 addition from
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L481

Cost at the pixel level for just `bilinear`:
https://github.com/tensorflow/tensorflow/blob/acab6abb0113ff26f17fd0bc0f7dee7af7c3b41f/tensorflow/core/kernels/crop_and_resize_op.cc#L488

Cost at the depth level for `bilinear`:
3 additions for `1 - y_lerp` and `1 - x_lerp`
4 additions for `+=`
6 multiplications

Cost at the depth level for `nearest`
1 addition and 1 cast.

Anything else?

wenxizhu replied:
Generally I agree. However I think there are 4 casts in `bilinear` depth level.

So that's our conclusion:
1. for `bilinear`:
    4 addition + depth * (7 addition + 6 multiplication + 4 cast)
2. for 'nearest':
    3 addition + depth * (1 addition + 1 cast)

penpornk replied:
Right. I forgot about the casts in bilinear. Thank you!

penpornk reviewed (commented):
Looks great! A question please.

wenxizhu commented:
@peppornk For the "additions count" problem, I already gave my best guess. But since I'm not the author of that code, I also don't know where my guess is correct. Your opinions?

wenxizhu commented:
@penpornk Any updates?

penpornk reviewed (commented):
So sorry for my late reply and thanks for your patience! Let's agree on the final cost and this PR should be done.

wenxizhu committed:
Update cost estimation for CropAndResizeBackpropImage:
1. bilinear:
   4 addtion + depth * (7 addtion + 6 multiplication + 4 cast);
2. nearst:
   3 addtion + depth * (1 addtion + 1 cast)

wenxizhu commented:
@penpornk Changes applied for cost estimation update. Please see my commit for details.

penpornk reviewed (approved):
Thank you again for the PR!

wenxizhu commented:
@penpornk You're welcome. And thank you for reivew this PR!

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -425,7 +426,7 @@ class CropAndResizeGradImageOp : public AsyncOpKernel {
 template <typename T>
 struct CropAndResizeBackpropImage<CPUDevice, T> {
-  bool operator()(const CPUDevice& d,
+  bool operator()(const OpKernelContext* context,
```
Could you please change the first parameter back to `CPUDevice`?
```suggestion
  bool operator()(const CPUDevice& d,
```

wenxizhu replied:
Hmm...we can do that. But how to handle the following line:
https://github.com/tensorflow/tensorflow/blob/5173d9d9ff0cfcb5e175b801c21b10e2e3704dc3/tensorflow/core/kernels/crop_and_resize_op.cc#L530

penpornk replied:
Oh, right. It has been a while I forgot what we need the context for. Then we'll need to fix the build errors instead.

penpornk reviewed (changes requested):
We are getting some build errors.

penpornk reviewed (changes requested):
Please fix the build errors. For example, in GPU Python 3 test [log](https://source.cloud.google.com/results/invocations/e3c983d5-6724-40ac-aab2-674609c058ba/log) line 5666:
`./tensorflow/core/kernels/crop_and_resize_op.h:41:8: note:   no known conversion for argument 1 from 'tensorflow::OpKernelContext* const' to 'const Eigen::GpuDevice&'1`

wenxizhu committed:
Make changes to "crop_and_resize_op.cc" to make it pass GPU build.

wenxizhu commented:
@penpornk Hi, changes made for GPU build. Let's see if it works.

penpornk reviewed (changes requested):
I don't think we need to add code for GPUDevice. Please remove them. I think we just need to update the function definition in the .h file to match the implementation, i.e., change `const Device& d` to `const OpKernelContext* context`.
https://github.com/tensorflow/tensorflow/blob/4b1ee1a7b1fb29fae15920d827a7ed3878f7c6af/tensorflow/core/kernels/crop_and_resize_op.h#L38-L46

wenxizhu committed:
Revert "Make changes to "crop_and_resize_op.cc" to make it pass GPU build."

This reverts commit c137aa421c96658e9010bdd1a47f114dec5f6005.

wenxizhu committed:
Change signature of CropAndResize() and CropAndResizeBackpropImage, from Device* to OpKernelContext*

penpornk commented:
@wenxizhu Thank you for the quick fix! Let's see if it passes now.

penpornk commented on tensorflow/core/kernels/crop_and_resize_op.cc:
```diff
@@ -198,9 +199,8 @@ class CropAndResizeOp : public AsyncOpKernel {
-struct CropAndResize<CPUDevice, T> {
-  bool operator()(const OpKernelContext* context,
+template <typename Device, typename T>
+bool CropAndResize<Device, T>::operator()(const OpKernelContext* context,
```
We don't need to change this from functor to normal functions.

penpornk reviewed (commented):
I just took a look at the fix. It changes functors to functions. I don't want to make unnecessary changes that aren't related to the errors, so let's stop here. I've fixed the errors (based on the first time this PR is pulled in) internally. It will take some time for other people to review it (since I can't review the changes I made), after that it should be merged.

penpornk commented:
The PR is merged in https://github.com/tensorflow/tensorflow/commit/ddf3966936b65186b46cafe52181e0c2af480fae. I'm closing this PR now. Thank you again for your contributions!
